File: uptime.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	ErrorUnableToReadFile = "couldn't read uptime file"
	ErrorUptimeNotFloat   = "uptime value is not decimal"
)

func getUptime() uint64 {
	bytes, err := os.ReadFile("/proc/uptime")
	if err != nil {
		log.Fatalln(ErrorUnableToReadFile)
	}

	contents := string(bytes[:])
	parts := strings.Split(contents, " ")

	float, err := strconv.ParseFloat(parts[0], 64)
	if err != nil {
		log.Fatalln(ErrorUptimeNotFloat)
	}

	return uint64(float)
}

func getUptimeMessage() string {
	seconds := getUptime()

	values := []struct {
		uint64
		string
	}{
		{seconds / 60 / 60 / 24, "day"},
		{seconds / 60 / 60 % 24, "hour"},
		{seconds / 60 % 60, "minute"},
		{seconds % 60, "second"},
	}

	var toReturn []string

	for _, value := range values {
		if value.uint64 == 0 {
			continue
		}

		if value.uint64 > 1 {
			value.string += "s"
		}

		toReturn = append(toReturn, fmt.Sprintf("%d %s", value.uint64, value.string))
	}

	return strings.Join(toReturn, " ")
}

File: uptime.py
#!/bin/env python

import json
import os
import sys
import time
from http.client import HTTPSConnection
from urllib.parse import urlparse

CONFIG_FILE = f'{os.environ['HOME']}/.config/house-monitor.conf'
REQUIRED_PROPERTIES = ['DEVICE_NAME', 'DISCORD_WEBHOOK']
UPTIME_FILE = '/proc/uptime'


def read_config():
    config = {}

    with open(CONFIG_FILE, 'r') as f:
        for line in f.readlines():
            (key, value) = line.strip().split('=')
            config[key] = value

    for prop in REQUIRED_PROPERTIES:
        if not prop in config.keys():
            print(f"Required property '{prop}' not found in config file")
            sys.exit(1)

    return config


def get_message():
    with open(UPTIME_FILE, 'r') as f:
        seconds = int(float(f.read().split(' ')[0]))

        units = [
            (seconds // 60 // 60 // 24, "day"),
            (seconds // 60 // 60 % 24, "hour"),
            (seconds // 60 % 60, "minute"),
            (seconds % 60, "second"),
        ]

        return ' '.join([
            f'{x} {y if x == 1 else f'{y}s'}'
            for (x, y) in units
            if x > 0
        ])


def send_message(webhook, message):
    url = urlparse(webhook)
    conn = HTTPSConnection(url.netloc)

    conn.request(
        'POST',
        url.path,
        json.dumps({'content': message}),
        {'Content-Type': 'application/json'}
    )

    conn.getresponse()


def main():
    timestamp = int(time.time())
    config = read_config()

    send_message(
        config['DISCORD_WEBHOOK'],
        f'**[<t:{timestamp}>][{config['DEVICE_NAME']}]** Uptime: {get_message()}'
    )


if __name__ == '__main__':
    main()
